File: articles/design-research-methods-contextual-inquiry-synthesis/c/theme_confidence_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ROWS 256
#define MAX_THEMES 64
#define MAX_LINE 512
#define MAX_TEXT 128

typedef struct {
    char theme[MAX_TEXT];
    int evidence_units;
    double total_strength;
    double total_risk;
    double confidence;
} ThemeAggregate;

int find_theme(ThemeAggregate *themes, int count, const char *theme) {
    for (int i = 0; i < count; i++) {
        if (strcmp(themes[i].theme, theme) == 0) {
            return i;
        }
    }
    return -1;
}

int compare_desc(const void *a, const void *b) {
    const ThemeAggregate *ta = (const ThemeAggregate *)a;
    const ThemeAggregate *tb = (const ThemeAggregate *)b;
    if (ta->confidence < tb->confidence) return 1;
    if (ta->confidence > tb->confidence) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/contextual_inquiry_evidence_units_raw.csv";
    FILE *file = fopen(path, "r");
    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    ThemeAggregate themes[MAX_THEMES];
    int theme_count = 0;
    char line[MAX_LINE];

    memset(themes, 0, sizeof(themes));
    fgets(line, sizeof(line), file);

    while (fgets(line, sizeof(line), file)) {
        char *fields[12];
        int field_count = 0;
        char *token = strtok(line, ",");

        while (token != NULL && field_count < 12) {
            fields[field_count++] = token;
            token = strtok(NULL, ",");
        }

        if (field_count < 7) {
            continue;
        }

        const char *theme = fields[3];
        double strength = atof(fields[5]);
        double risk = atof(fields[6]);

        int idx = find_theme(themes, theme_count, theme);
        if (idx < 0 && theme_count < MAX_THEMES) {
            idx = theme_count++;
            strncpy(themes[idx].theme, theme, MAX_TEXT - 1);
            themes[idx].theme[MAX_TEXT - 1] = '\0';
        }

        if (idx >= 0) {
            themes[idx].evidence_units += 1;
            themes[idx].total_strength += strength;
            themes[idx].total_risk += risk;
        }
    }

    fclose(file);

    for (int i = 0; i < theme_count; i++) {
        double mean_strength = themes[i].total_strength / themes[i].evidence_units;
        double mean_risk = themes[i].total_risk / themes[i].evidence_units;
        themes[i].confidence = 0.70 * mean_strength - 0.30 * mean_risk;
    }

    qsort(themes, theme_count, sizeof(ThemeAggregate), compare_desc);

    printf("rank,theme,evidence_units,synthesis_confidence\n");
    for (int i = 0; i < theme_count; i++) {
        printf("%d,%s,%d,%.6f\n", i + 1, themes[i].theme, themes[i].evidence_units, themes[i].confidence);
    }

    return 0;
}

File: articles/design-research-methods-contextual-inquiry-synthesis/go/theme_confidence_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type ThemeAggregate struct {
	Theme          string
	EvidenceUnits  int
	TotalStrength  float64
	TotalRisk      float64
	Confidence     float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func main() {
	path := "../data/raw/contextual_inquiry_evidence_units_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	aggregates := map[string]*ThemeAggregate{}

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 7 {
			continue
		}

		theme := row[3]
		strength := parseFloat(row[5])
		risk := parseFloat(row[6])

		if _, ok := aggregates[theme]; !ok {
			aggregates[theme] = &ThemeAggregate{Theme: theme}
		}

		aggregates[theme].EvidenceUnits++
		aggregates[theme].TotalStrength += strength
		aggregates[theme].TotalRisk += risk
	}

	var out []ThemeAggregate
	for _, agg := range aggregates {
		meanStrength := agg.TotalStrength / float64(agg.EvidenceUnits)
		meanRisk := agg.TotalRisk / float64(agg.EvidenceUnits)
		agg.Confidence = 0.70*meanStrength - 0.30*meanRisk
		out = append(out, *agg)
	}

	sort.Slice(out, func(i, j int) bool {
		return out[i].Confidence > out[j].Confidence
	})

	fmt.Println("rank,theme,evidence_units,synthesis_confidence")
	for i, row := range out {
		fmt.Printf("%d,%s,%d,%.6f\n", i+1, row.Theme, row.EvidenceUnits, row.Confidence)
	}
}
